Migrate to the Stream Deck SDK and HomeAssistant WebSocket Client

tlusk opened the pull request:
~~This is currently in a WIP state.~~ This should now be feature complete.

* The communication between the property inspector and the running plugin is mostly complete.
  * Both global and standard settings are saving and loading and rendering in the UI
  * The property inspector no longer has a connection to HomeAssistant at all. All communication is done via events to and from the plugin side. This change makes the property inspector side really just a view to render the settings and nothing else.
* ~~Button presses to trigger services are working (long presses, and dials are not yet implemented).~~
* ~~Rendering *anything* on the button itself is not yet implemented.~~
* *Both* vite and rollup are being used to build the project
  * I originally played around with removing vue entirely and using the https://sdpi-components.dev/ as recommended in the Stream Deck SDK documentation, but doing so turned into a much bigger lift and can be left for a future consideration.
  * I spent far too much time trying to either get rollup to build a vue project, or vite to build a node application and ended up being unsuccessful.  We can revisit this in the future but for now it's working using both tools in series.
* ~~The plugin name was changed to include a `2` just for my local development so I can have both the latest published version and this WIP installed side by side to compare.  This will be reverted before this MR is marked as complete.~~

cgiesche commented:
Thank you very much for your work! I was in the hospital for a while, which is why I haven't had a chance to look at your changes yet. 🩻 I will review them in the coming days and simultaneously try them out on my computer. :)

cgiesche commented:
For some entities, the label rendering seems to be a bit broken:
![grafik](https://github.com/user-attachments/assets/a2ceaf1e-6718-42a7-a473-73500bbea08a)
(weather)

![grafik](https://github.com/user-attachments/assets/015d84e3-c116-4f53-8ffb-b666895facfd)
(temperature / humidity)

Besides that, everything seems to work as expected.

tlusk commented:
@cgiesche do you have those entities configured with a custom template? Just trying to figure out what might be different with those ones.

cgiesche commented:
Maybe you have the entity "Forecase Home". I think this is always available.

tlusk commented on src/render/svgUtils.ts:
```diff
@@ -0,0 +1,162 @@
+  }
+
+  private drawText(text: string, lineNr: number) {
+    const escapedText = encodeURIComponent(text)
```
@cgiesche It looks like this line is the one causing a problem... but it's not new 😕.

Removing it makes it render correctly.  Maybe just a difference between the old and new SVG libraries.

tlusk commented:
@cgiesche I added unit tests into the project, well a single test so far. The test I added tests the svg rendering logic with text that contains the ° degree symbol which was causing the bug you caught.

I'll add some more tests around the svg rendering when I get a chance.

cgiesche commented:
> well a single test so far

Unit tests are for the weak. No, seriously, I did it the way projects typically go in real life. When you're short on time, tests are the first thing to go out the window. :)

I even remember URL-encoding the values back then, because the SVG library would completely crash if you gave it non-encoded characters. Apparently, that's been fixed now. 

Anyway, it works now 👍

tlusk commented:
@cgiesche I finished up adding additional unit tests around the SVG rendering. I might try to add some other tests for other components but that can come at any point, it's not a blocker for getting this merged in.

tlusk commented:
@cgiesche try removing this commit specifically: https://github.com/cgiesche/streamdeck-homeassistant/commit/f78a71e91decd276e4281f418949f62ae6fb1015

Also getting logs from the plug-in itself would be great, all the logs I've seen posted look to be from home assistant itself?

tlusk commented:
@cgiesche the more I think about this the more I'm confident that commit is the issue. It seems like if one entity is set in the filter that it has an issue with the entire subscription will fail and you'll get no status for anything. 

Remove just that commit and it should fix the issue.

tlusk commented:
@cgiesche I'm also concerned about the duplicate domain some people seem to have. Maybe try to get someone who posted the log with the duplicate domain to share their home assistant configuration file for one of the pages they have configured so you can take a look at the json and see if something is saved wrong in their config.

tlusk commented:
@cgiesche Actually it's possible it's that they have data saved on the tiles in an older settings version (v2/v3 etc...) and something is wrong with the migration.

I wrote unit tests explicitly around that though so... let's try to get a backup of one of their profiles to see how their settings looks.

Kleberstoff commented:
Hello,

hopefully this is ok for me to post here, if not yell at me please.

i have tried playing around with both version, for me not working 3.6 and working 3.5/3.7.
Logs from the plugin side seem not very helpfulm I have only included 3.6.

What I've done is, remove the plugin, restart Stream Deck, install the Version and try.

OS: Windows 11 Pro, 24H2, Build 26100.3775
StreamDeck: 6.8.1 (21263)
Home Assistant: 
Core 2025.4.3
Supervisor 2025.04.0
Operating System 15.2
Frontend 20250411.0

The connection only seems to be working sometimes(?, i had like a 10% chance of it connecting) if I use my external domain.
Which could be because my internal domain is using a internal root CA.

I had some connects with not showing the entities last time I tried, but I couldn't get it right now, but if it connected it actions worked, states didn't.

Plugin logs:
[de.perdoctus.streamdeck.homeassistant.0.log](https://github.com/user-attachments/files/19857683/de.perdoctus.streamdeck.homeassistant.0.log)
[de.perdoctus.streamdeck.homeassistant.1.log](https://github.com/user-attachments/files/19857684/de.perdoctus.streamdeck.homeassistant.1.log)
[de.perdoctus.streamdeck.homeassistant.2.log](https://github.com/user-attachments/files/19857685/de.perdoctus.streamdeck.homeassistant.2.log)
[de.perdoctus.streamdeck.homeassistant.3.log](https://github.com/user-attachments/files/19857686/de.perdoctus.streamdeck.homeassistant.3.log)
[de.perdoctus.streamdeck.homeassistant.4.log](https://github.com/user-attachments/files/19857687/de.perdoctus.streamdeck.homeassistant.4.log)
[de.perdoctus.streamdeck.homeassistant.5.log](https://github.com/user-attachments/files/19857688/de.perdoctus.streamdeck.homeassistant.5.log)

StreamDeck logs:
[StreamDeck0.log](https://github.com/user-attachments/files/19857691/StreamDeck0.log)

StreamDeck Profile(.streamDeckProfile):
[Profile.txt](https://github.com/user-attachments/files/19857725/Profile.txt)

StreamDeck backup(.streamDeckProfilesBackup):
[Stream Deck - 23-04-2025 - 00-32.txt](https://github.com/user-attachments/files/19857729/Stream.Deck.-.23-04-2025.-.00-32.txt)

I changed the extension to .txt so I can upload them to github, hope thats fine.

I hope its somewhat useful, if I can provide more detailed logs or be helpful in an other way, let me know.

Sorry if the formatting isn't great, I rarely use github.

Best regards,
Leon

cgiesche commented:
Hey @tlusk,

no worries at all and thanks so much for your support! Don’t feel the need to apologize – I barely find time for the project myself, and with an expected little one on the way soon it will probably get even trickier over here.   

Regarding the fix, I’m relying on removing the (possible) duplicate domains as you suggested. As for subscribing only to needed entities, that’s been working perfectly for me for months now, even across different Home Assistant versions.   

Whether it’s worth adding a toggle for that?

Thanks again for your help, really appreciated!
